### DISCUSSION TITLE
form.getInputProps(path) is not type safe

### DISCUSSION THREAD

**Discussion by testerez** [mantinedev/mantine#2502] [Q&A]
Hello,
I noticed that `form.getInputProps(path)` can be called with any `path`. Since I'm setting a `<Values>` generic on `useForm`, I was expecting that only valid paths would be accepted.
Is there any way to enforce this? Or maybe someone can explain why it was designed that way?

**Comment by rtivital** [comment]
It is designed this way to accept nested paths, for example `user.data.email` – https://mantine.dev/form/nested/

**Comment by testerez** [comment]
Thanks for the quick answer @rtivital ,

It looks like dot notation could be fairly easily inferred from `Values`:

```ts
type BreakDownObject<O, R = void> = {
  [K in keyof O as string]: K extends string
    ? R extends string
      ? ObjectDotNotation<O[K], `${R}.${K}`>
      : ObjectDotNotation<O[K], K>
    : never
}


type ObjectDotNotation<O, R = void> = O extends string
  ? R extends string
    ? R
    : never
  : BreakDownObject<O, R>[keyof BreakDownObject<O, R>]

type Values = {
  name: string
  address: { street: string }
}
type FormPath = ObjectDotNotation<Values>
const formPath: ValuesPath = 'address.street' // type FormPath = "name" | "address.street"
```

Source: https://stackoverflow.com/a/71442049/508194

Edit: this solution would require a bit more work to support arrays.

**Reply by rtivital** [reply]
Thanks for sharing, I'll consider adding it

**Reply by testerez** [reply]
For info @rtivital , I've found [this version](https://www.typescriptlang.org/play?ts=4.1.0-dev.20200921#code/C4TwDgpgBACghsAFgSQLZgDYB4AqAaKAaSggA9gIA7AEwGcoBrCEAewDMocA+KAXgCgoREuSp0otYACcAlpQDmgqAH5OAbUIBdERRr0AShADGLKdSyTZCgnEoguSoapwbtZXeICCUqXBAAZGSYsW3tHIRVhAB8oAAMAEgBvQgBfADok+CQ0TFxXAgBRUiMMAFdqCCwmVg4XLQJq9ihQtU0uLhTY8KEALmi4pNSMxKyUdGw6zQbmJsmOroioPsIlPsoIADcIKQBufn5QSFgERFweXmPs8dxpms4eGMbavYPwaFGANTgyyvxYHTE9FGZ3OShgAL0A0ScjY2yI6SSMLhhkknSUqmI7kBjBmtXCqhRwAh4mBc26kU+31Kv3yUEJDkWvSg6y2UnCa0221W-yxkKenHx6hgmnZzM5u32FRKcCk0DYpUoRmAMhYlCg8ggwBuPNEkNJ7QAFCwAEYAKz6fzAJz6MAAlDaTl8ftqYFwXiZKJIoCbTcYiRdEko2DIpJIAHJwVAQPoAIgAIjIIPIWDG8EoMHBw5Ho1AYwAJOAAL1TSjgGr6AGYAAxpoRgKQsX1K2h9NThRLM7OxwxwBgyYCpqAe6QyY2lYCmFtQABsAA4oClaxEO5Qu7mAMKqyQICCD4eyMcT0N9ACMACYF0uRSlmvQPZJ3VuiRsqdALhrgEazX6CDH642-VoNIY1tPYgA) that works a bit better for me

**Reply by testerez** [reply]
I modified a bit more. Here is the [playground](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBACghsAFgSQLZgDYB4AqAaKAaSggA9gIA7AEwGcoBrCEAewDMocA+KAXgCgoREuSp0oAClrAATgEtKAcygAfKJQCuqAEYQZASkFQA-JwDahALoiKNegCUIAYxYzqWafKUE4lEFyMhUxwLazJbcQBBGRk4EAAZOSYsX39AoRNhNQADABIAb0IAXwA6AvgkNExcUIJNHT0uIuz0oQAuLKg8wtLyhBR0bBCrAiZWDmHLJpaMqA7CIw7KCAA3PQBufn4yMFdgKFBIKAB5bQArZ2AKxFweXlh+qqHR5nZOHjUxt5xNrYB6P6cCDSNpbQ7QACqtD0fCg+SMlDgqAgHU8CkUeCMcGo1BkwNoHXhswwCmBhNaUE8EAgwFRsnRmNmQipNIAclpdDIlhy9OkimZLIyMgAvORgOleDFGIr8Iq-FyUaSU2TU4Dshpck7nS7XLBQxqwgBE2Nx+JKJOWtBKACYSiy1TyZIaoACoIBQcn4CqVbBYLA6pwuTiu-T10JkdygxpxeNoVotwJtJR9LGdrsAMuRQPGB4D0BQrOAk6hQMD9KBweg7S4Qaj8IA)

**Reply by DenisBessa** [reply]
@testerez any chance you make a PR implementing this funcionality? It would be really useful

**Reply by testerez** [reply]
Hey @DenisBessa , time is hard to find and I'd need to have clear buyin from the Maintainer since such a change would certainly cause typing errors for many users.

**Comment by DenisBessa** [comment]
This would be a very nice feature!

**Comment by jrmdayn** [comment]
This [repo](https://github.com/anthonyjoeseph/spectacles-ts) could be helpful for that.. or at least give ideas
